revoke/revoke.go

```diff
@@ -123,48 +263,90 @@ func getIssuer(cert *x509.Certificate) *x509.Certificate {
 	}
 
 	return issuer
-
 }
 
-// check a cert against a specific CRL. Returns the same bool pair
-// as revCheck.
-func certIsRevokedCRL(cert *x509.Certificate, url string) (revoked, ok bool) {
-	crl, ok := CRLSet[url]
+// checks whether CRL in memory is valid
+func (r *Revoke) isInMemoryCRLValid(key string) bool {
+	crl, ok := r.crl.set[key]
 	if ok && crl == nil {
 		ok = false
-		delete(CRLSet, url)
+		delete(r.crl.set, key)
+	} else if crl == nil {
+		delete(r.crl.set, key)
+		return false
 	}
 
-	var shouldFetchCRL = true
-	if ok {
-		if !crl.HasExpired(time.Now()) {
-			shouldFetchCRL = false
-		}
+	if ok && !crl.HasExpired(time.Now()) {
+		return true
+	}
+
+	return false
+}
+
+// fetchLocalCRL reads CRL from the local filesystem
+// force flag allows you to update the CRL
+func (r *Revoke) fetchLocalCRL(newLocalCRL string) (*pkix.CertificateList, error) {
+	if _, err := os.Stat(newLocalCRL); err != nil {
+		return nil, fmt.Errorf("failed to read local CRL path: %v", err)
 	}
 
-	issuer := getIssuer(cert)
+	tmp, err := ioutil.ReadFile(newLocalCRL)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read local CRL path: %v", err)
+	}
+
+	crl, err := x509.ParseCRL(tmp)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse local CRL file: %v", err)
+	}
+
+	if crl == nil {
+		return nil, fmt.Errorf("Local CRL is nil")
```

Comment:
ParseCRL returns ParseDERCRL (https://golang.org/src/crypto/x509/x509.go?s=51887:51948#L1647). ParseDERCRL returns crl=nil only when there is an error. I don't know whether new golang releases would return something else. That is why I added extra check. If you think I have to remove this check, so ok.